controllers/stargate_controller.go

```diff
@@ -122,88 +122,120 @@ func (r *StargateReconciler) Reconcile(ctx context.Context, req ctrl.Request) (c
+			if actualDeploymentHash, found := actualDeployment.Annotations[api.ResourceHashAnnotation]; !found || actualDeploymentHash != desiredDeploymentHash {
+				logger.Info("Updating Stargate Deployment", "Deployment", deploymentKey)
+				resourceVersion := actualDeployment.GetResourceVersion()
+				desiredDeployment.DeepCopyInto(&actualDeployment)
+				actualDeployment.SetResourceVersion(resourceVersion)
+				// Set Stargate instance as the owner and controller
+				if err := ctrl.SetControllerReference(stargate, &actualDeployment, r.Scheme); err != nil {
+					logger.Error(err, "Failed to set controller reference on updated Stargate Deployment", "Deployment", deploymentKey)
+					return ctrl.Result{}, err
+				} else if err := r.Update(ctx, &actualDeployment); err != nil {
+					logger.Error(err, "Failed to update Stargate Deployment", "Deployment", deploymentKey)
+					return ctrl.Result{}, err
+				} else {
+					logger.Info("Stargate Deployment updated successfully", "Deployment", deploymentKey)
+					return ctrl.Result{RequeueAfter: time.Minute}, nil
+				}
+			}
+			delete(desiredDeployments, actualDeployment.Name)
+			replicas += actualDeployment.Status.Replicas
+			readyReplicas += actualDeployment.Status.ReadyReplicas
+			updatedReplicas += actualDeployment.Status.UpdatedReplicas
+			availableReplicas += actualDeployment.Status.AvailableReplicas
 		}
 	}
 
-	// Check if the deployment needs to be updated
-	desiredDeploymentHash := desiredDeployment.Annotations[api.ResourceHashAnnotation]
-	if actualDeploymentHash, found := actualDeployment.Annotations[api.ResourceHashAnnotation]; !found || actualDeploymentHash != desiredDeploymentHash {
-		logger.Info("Updating Stargate Deployment", "Deployment", deploymentKey)
-		resourceVersion := actualDeployment.GetResourceVersion()
-		desiredDeployment.DeepCopyInto(actualDeployment)
-		actualDeployment.SetResourceVersion(resourceVersion)
+	for _, desiredDeployment := range desiredDeployments {
+		// Deployment does not exist yet: create a new one
+		deploymentKey := client.ObjectKey{Namespace: req.Namespace, Name: desiredDeployment.Name}
+		logger.Info("Stargate Deployment not found, creating a new one", "Deployment", deploymentKey)
 		// Set Stargate instance as the owner and controller
-		if err := ctrl.SetControllerReference(stargate, actualDeployment, r.Scheme); err != nil {
-			logger.Error(err, "Failed to set controller reference on updated Stargate Deployment", "Deployment", deploymentKey)
-			return ctrl.Result{}, err
-		} else if err := r.Update(ctx, actualDeployment); err != nil {
-			logger.Error(err, "Failed to update Stargate Deployment", "Deployment", deploymentKey)
+		if err := ctrl.SetControllerReference(stargate, &desiredDeployment, r.Scheme); err != nil {
+			logger.Error(err, "Failed to set controller reference on new Stargate Deployment", "Deployment", deploymentKey)
 			return ctrl.Result{}, err
+		} else if err := r.Create(ctx, &desiredDeployment); err != nil {
```

Comment:
Good idea, will do.

Reply:
I like the overall idea, but how would you deal with the aggregated errors? Is it ok to return, say, the first one?

Reply:
>how would you deal with the aggregated errors? 

I think we would need a custom error type whose `Error()` method provides a summary of all the errors. As each error occurs we should log it as well.

Reply:
Thinking a bit more about this idea, I think It might not be a good idea. Looking at cass-operator code, every error encountered along the (very long) reconciliation process stops the reconcile loop immediately – even when inside a  range loop, for example in `CheckRackStoppedState`.

I would suggest that, in a follow-up PR, we implement something similar to cass-operator's `ReconcileResult` interface. This would greatly simplify the code as we would be able to easily split the reconciliation logic into many small, independent functions.

Once this new utility is available, then maybe we could add support for accumulating multiple errors until we reach the next checkpoint (a point in the reconcile loop where we must stop if there is any error so far). Does that make sense to you?

Reply:
@adutra can you clarify what you think might be a bad idea? I am not sure if you are referring to the cass-operator process of failing fast or trying to fail later.

I am certainly in favor of using `ReconcileResult` or something similar if it will help simplify the reconciliation logic.

And I am definitely fine with making the change in a follow-up PR as it is more of a enhancement than anything else.

Reply:
I'm sorry, what I meant is that it would be a bad idea _if done within this PR_. I do think your idea is good, but I would have to introduce at least some support for accumulating errors and merging `ctrl.Result` instances. What I meant to say is just that it would be better to do it in a follow-up PR. I'd be happy to bootstrap this work as soon as this PR is merged.